### PR TITLE
Change type of date on Todo task to Date

### LINKED ISSUE #8405: API: dates may be invalid
[//]: # (Fill out relevant information - UUID is found in Settings -> API)
General Info
  * UUID: b0413351-405f-416f-8787-947ec1c85199
  * Browser: Safari
  * OS: MacOSX 10.11 (el capitain)

### Description
[//]: # (Describe bug in detail here. Include pictures if helpful.)
When submitting new tasks through the tasks API, anything can go into many of the date params.

For example if I POST the following task: 
{type":"todo","text":"Practice Task","date":"ABCDE"}

Then the Due Date on the task listed on the Habitica site will be "ABCDE"

The only date that seems to be forced to be correct is "startDate" for "reminders"

#### Console Errors
[//]: # (Include any JavaScript console errors here.)

### Notes
If you want to work on this this PR https://github.com/HabitRPG/habitica/pull/10625 has some useful code but the validation should be moved to the model. A migration to remove existing invalid dates is also going to be needed

### PR DESCRIPTION
**Pull request by sprusr**
[//]: # (Note: See http://habitica.wikia.com/wiki/Using_Habitica_Git#Pull_Request for more info)

[//]: # (Put Issue # or URL here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes #8405 

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)

Was previously set to String, with no validation. Schema now checks for
correct date, and stores as a Date. Also includes a migration to
convert valid date strings to Dates in the database, and remove invalid
strings.

[//]: # (Put User ID in here - found in Settings -> API)

----
UUID: db711097-c5d0-4d25-8e5e-d955da4b79fd

### CODE REVIEW

File: migrations/20170115_api_date_validation.js
Comment (khipkin) [inline migrations/20170115_api_date_validation.js]: Suggestion: if this is just for testing the script, put a "TODO" comment so you remember to remove it.

File: migrations/20170115_api_date_validation.js
Comment (khipkin) [inline migrations/20170115_api_date_validation.js]: function isValidDate(d) {
        return Object.prototype.toString.call(d) === "[object Date]" && !isNaN(d.getTime());
    }

File: migrations/20170115_api_date_validation.js
Comment (sprusr) [reply migrations/20170115_api_date_validation.js]: Ah yes, well spotted. This was indeed for testing purposes, will remove. Thanks.

**Comment by sprusr** [comment]
For the sake of cleanliness, I've amended that to the previous commit. Should be ready to go now @khipkin

**Comment by khipkin** [comment]
You should probably add a test that non-date values now fail.

**Comment by librarianmage** [comment]
Is this related to #7715? If so, would you be able to fix that here as well? I've completely forgotten to work on it, and this seems to be related/the same problem.

File: migrations/20170115_api_date_validation.js
Comment (librarianmage) [inline migrations/20170115_api_date_validation.js]: Does this need to be changed?

**Comment by khipkin** [comment]
@MathWhiz #7715 seems to include UI changes as well (disallow saving tasks with invalid start dates).

**Comment by sprusr** [comment]
@MathWhiz does seem to be out of the scope of this fix, but thanks for the heads up about the migration name

**Comment by librarianmage** [comment]
@sprusr Ok. I will wait until your PR gets merged in to start working on it.

**Comment by Alys** [comment]
@sprusr I like this change and I love that you did the migration script!

I think though that we need to delay merging it because it could count as a breaking change for the API. There might be third-party tools using that rely on the date field being free-text and/or that rely on a specific date format in the users' tasks. I think we definitely should make this change though - it's clearly an improvement. We can advertise the change and the data migration in the coders guilds in Habitica and in our dev blog and then wait a while before merging.

@paglias @crookedneighbor Tell us if you disagree!

**Comment by paglias** [comment]
I'm in favor of the change, I consider it a bug fix instead of a breaking change

**Comment by sprusr** [comment]
@khipkin got round to adding a test. Not very experienced with writing tests, so please let me know if it can be improved (edit: oops I did that wrong, fixed now).

@Alys @paglias I would think it would depend on what the API docs say about setting the due date. I just tried to have a look, but didn't find [the docs](https://habitica.com/apidoc/#api-Task-CreateUserTasks) very descriptive. Is the schema for a todo task documented somewhere? Otherwise, I reckon those docs could be an area of improvement and might have a look into that myself 😁

But yeah, merge it when you feel appropriate. Thanks to all for the guidance in contributing to this.

File: migrations/20170115_api_date_validation.js
Comment (paglias) [reply migrations/20170115_api_date_validation.js]: yes

File: migrations/20170115_api_date_validation.js
Comment (paglias) [inline migrations/20170115_api_date_validation.js]: I don't think this is going to work as $nor matches docs whose date type is not a date AND where the date fields doesn't exist. And i don't think is necessary as all date fields are going to be strings if they exist. You can simply match `date: {$exists: true}`

File: migrations/20170115_api_date_validation.js
Comment (paglias) [inline migrations/20170115_api_date_validation.js]: the migration field doesn't exits on tasks, let's not set it

File: migrations/20170115_api_date_validation.js
Comment (paglias) [inline migrations/20170115_api_date_validation.js]: same here

File: migrations/20170115_api_date_validation.js
Comment (paglias) [inline migrations/20170115_api_date_validation.js]: the migration should be updated to use `monk` instead of mongoskin that doesn't support anymore our database version, see https://github.com/HabitRPG/habitica/blob/develop/migrations/20170120_missing_incentive.js#L10

**Review by paglias — CHANGES_REQUESTED** [review]

**Comment by sprusr** [comment]
@paglias fingers crossed, that should sort everything

File: migrations/20170115_api_date_validation.js
Comment (paglias) [inline migrations/20170115_api_date_validation.js:41]: what is this for? To make sure only one task is processed at a certain time? There are more than 20 milions tasks that have this field so even with a low processing time of 5ms per task (and with latency it's going to be a lot more) it'll take more than a day to run the migration. I think we want multiple tasks processed at the same time. For example fetch 100 tasks, process them while we fetch the next 10 and when the first 10 have been processed repeat the process

File: migrations/20170115_api_date_validation.js
Comment (paglias) [inline migrations/20170115_api_date_validation.js:43]: this should be returned in order for errors to be catched by the `.catch` below

**Review by paglias — CHANGES_REQUESTED** [review]

**Comment by SabreCat** [comment]
@sprusr, this PR has been awaiting your input for a while. Are you still planning to address the comments here?

**Comment by SabreCat** [comment]
No response to our query, so I'm closing this and reopening the source issue as help-welcome.

**Pull request closed without merging**
